ttnn/cpp/ttnn/operations/matmul/matmul_pybind.hpp

```diff
@@ -136,109 +147,201 @@ void py_module(py::module& module) {
+        >>> tensor2 = ttnn.to_device(ttnn.from_torch(torch.randn((32), dtype=torch.bfloat16)), device)
+        >>> output = tensor1 @ tensor2
+        >>> print(output.shape)
+        [32]
+        >>> # matrix x vector
+        >>> tensor1 = ttnn.to_device(ttnn.from_torch(torch.randn((64, 32), dtype=torch.bfloat16)), device)
+        >>> tensor2 = ttnn.to_device(ttnn.from_torch(torch.randn((32), dtype=torch.bfloat16)), device)
+        >>> output = tensor1 @ tensor2
+        >>> print(output.shape)
+        [64, 1]
+        >>> # batched matrix x broadcasted vector
+        >>> tensor1 = ttnn.to_device(ttnn.from_torch(torch.randn((10, 64, 32), dtype=torch.bfloat16)), device)
+        >>> tensor2 = ttnn.to_device(ttnn.from_torch(torch.randn((32), dtype=torch.bfloat16)), device)
+        >>> output = tensor1 @ tensor2
+        >>> print(output.shape)
+        [10, 64, 1]
+        >>> # batched matrix x batched matrix
+        >>> tensor1 = ttnn.to_device(ttnn.from_torch(torch.randn((10, 64, 32), dtype=torch.bfloat16)), device)
+        >>> tensor2 = ttnn.to_device(ttnn.from_torch(torch.randn((10, 32, 128), dtype=torch.bfloat16)), device)
+        >>> output = tensor1 @ tensor2
+        >>> print(output.shape)
+        [10, 64, 128]
+        >>> # batched matrix x broadcasted matrix
+        >>> tensor1 = ttnn.to_device(ttnn.from_torch(torch.randn((10, 64, 32), dtype=torch.bfloat16)), device)
+        >>> tensor2 = ttnn.to_device(ttnn.from_torch(torch.randn((32, 128), dtype=torch.bfloat16)), device)
+        >>> output = tensor1 @ tensor2
+        >>> print(output.shape)
+        [10, 64, 128]
+        )doc",
+        ttnn::pybind_overload_t{
+            [](decltype(::ttnn::matmul)& self,
+               const ttnn::Tensor& input_tensor_a,
+               const ttnn::Tensor& input_tensor_b,
+               const bool transpose_a,
+               const bool transpose_b,
+               const ttnn::MemoryConfig& memory_config,
+               const std::optional<const DataType> dtype,
+               const std::optional<const MatmulProgramConfig> program_config,
+               const std::optional<const std::string>& activation,
+               const std::optional<const DeviceComputeKernelConfig> compute_kernel_config,
+               const std::optional<const ttnn::CoreGrid> core_grid) -> ttnn::Tensor {
+                return self(
+                    input_tensor_a,
+                    input_tensor_b,
+                    transpose_a,
+                    transpose_b,
                     memory_config,
                     dtype,
+                    program_config,
+                    activation,
                     compute_kernel_config,
-                    /*untilize_out=*/false,
-                    user_core_coord,
-                    get_fused_activation(activation),
-                    user_run_batched,
+                    core_grid);
+            },
```

Comment:
just fyi, you don't need this if the set of arguments bound to python is exactly the same as in invoke.
This is only needed if python api differs from c++

Reply:
Could you please elaborate what I don't need?

The pybind_overload_t ?

Reply:
I am not sure if you need it. 
It looks like bound arguments and args in invoke are the same and your lambda just proxies args - same args, same order.  
If thats truly the case, then you can remove it

Reply:
`pybind_overload_t` can be omitted if you only have a single (non-overloaded) `invoke` method.

If `invoke` is overloaded, then `pybind_overload_t` must be used for every overload that should be exposed to Python.

If `invoke` is not overloaded, then `bind_registered_operation` can take `pybind_arguments_t` as it's 4 argument.